sklearn/svm/base.py

```diff
@@ -168,8 +168,32 @@ def fit(self, X, y, sample_weight=None):
                              "boolean masks (use `indices=True` in CV)."
                              % (sample_weight.shape, X.shape))
 
-        if self.gamma == 'auto':
-            self._gamma = 1.0 / X.shape[1]
+        if self.gamma in ('scale', 'auto', 'auto_deprecated'):
+            if isinstance(X, sp.csr_matrix):
+                # std = sqrt(E[X^2] - E[X]^2)
+                X_std = np.sqrt((X.multiply(X)).mean() - (X.mean())**2)
+            else:
+                X_std = X.std()
+            if self.gamma == 'scale':
+                if X_std != 0:
+                    self._gamma = 1.0 / (X.shape[1] * X_std)
+                else:
+                    self._gamma = 1.0
+            else:
+                if (self.gamma == 'auto_deprecated' and
+                        self.kernel not in ('linear', 'precomputed') and
```

Comment:
This got lost somewho. I would rewrite this a bit for better readability:
```py
kernel_uses_gamma = self.kernel not in ('linear', 'precomputed') and not callable(self.kernel)
if (kernel_uses_gamma and not np.isclose(X_std, 1.0)):
    ...
```

Reply:
That would look neat.